### PR TITLE
Rewritten ensemble

### PR DESCRIPTION
**Pull request by mundya**
mem-sharing-ensemble
- [x] Profile compute
- [x] Profile network usage (should prove less bursty and less overall than current implementation)

### CODE REVIEW

File: spinnaker_components/ensemble/ensemble.c [hunk 1]
```diff
@@ -0,0 +1,448 @@
...
+  spin1_set_timer_tick(ensemble.parameters.machine_timestep);
+  spin1_callback_on(TIMER_TICK, timer_tick, 2);
+  spin1_callback_on(DMA_TRANSFER_DONE, dma_complete, 0);
+  spin1_callback_on(MCPL_PACKET_RECEIVED, mcpl_received, -1);
+  // --------------------------------------------------------------------------
+
+  // --------------------------------------------------------------------------
+  // Run the simulation
+  while (true)
+  {
```
Comment (mundya) [inline spinnaker_components/ensemble/ensemble.c:437]: Bug: should reset recording buffers here

**Comment by mundya** [comment]
After using CLZ to improve the decoding (top stack = decoding, middle stack = neuron update, bottom stack = input filtering).
plot_with_clz2

Compare with https://github.com/project-rig/nengo_spinnaker/pull/58#issuecomment-132252073

(Without using CLZ):
plot_with_destructive_shift

**Comment by mundya** [comment]
Note: these changes push ensembles much more toward being memory bound (e.g., a 800 neuron ensemble receiving and transmitting 16-d values will now be allocated to have 50 neurons per core as it hits memory boundedness far faster than compute boundedness [previously it would have been allocated 2x400 neuron cores]).
- [ ] ~~Page through encoders and decoders in SDRAM?~~

**Comment by mundya** [comment]
Correction: the memory calculation used when partitioning is wrong.

**Comment by mundya** [comment]
Confirmation that the SDRAM transfer and synchronisation aren't too horrible (note that the more data transferred [larger neuron count] and more synchronisation [more cores] delay the start of the decode and transmit step; and increase the variance in the start time of the neuron update and decode steps):

2 cores: stack_parallel_800_16

4 cores: stack_parallel_1600_16

8 cores: stack_parallel_3200_16

**Comment by mundya** [comment]
Also, the start times and durations do appear to be drawn from normal distributions:

dist_parallel_800_16

**Comment by mundya** [comment]
Some network results. I need to think a bit more (and/or talk to @mossblaser) about the best way to display + interpret these, but here goes.

## Experiment

An 800 neuron Ensemble representing a 16-d value with no input was probed. In `master` this instantiates three cores to simulate the ensemble, each of which transmits 16 packets per timestep. In this branch this instantiates a number of cores (n) which each transmit 16/n packets per timestep.

Using a script provided by @mossblaser (`sub_ms_packet_counters.py`) the router counters were polled to determine how many packets were routed per 5us samples for 10ms. The time-series of these results are shown below. (_Note_: the shifting on the time axis is because starting the packet counters at the exact same time for each experiment is rather hard):

time_series

As expected, `new-ensemble` sends far fewer packets overall than `master`; while still being bursty it is less bursty than `master` (the peaks are lower).

The following shows the number of times that a number of packets were routed in the same 5us sample (the same data as above, different view). Larger bars to the left of the chart are better (as this indicates lower packet injection rates):

injection_rates

**Conclusion**: as designed, `new-ensemble` sends far fewer packets overall than the implementation of Ensembles in `master`, and also does a better job of spreading transmission out over time.  This seems like a very good return for a slight extra cost in cycles.

_Note_: 800 neurons, 16-D is a common ensemble size in Spaun, so `new-ensemble` should make a huge difference to network performance when running models of Spaun-scale.

**Comment by mundya** [comment]
Final bit of profiling:
Input filtering cost in cycles ~= 40 \* Input dimensions + 131
Neuron update cost ~= (10 \* dimensions + 59) \* number of neurons
Decode and transmit cost ~= (3 \* number of neurons + 234) \* output dimensions

These numbers are fairly similar to what we had before apart from the decode stage. This is as expected. Any thoughts @neworderofjamie?

**Comment by mundya** [comment]
Now based on #94!

**Comment by mundya** [comment]
Hmm... something in that rebase went awry. I'll debug tomorrow.

**Comment by mundya** [comment]
And fixed. One character can screw things up quite a lot if it's a "=" in a ">=".

**Comment by mossblaser** [comment]
Very nice plots! Looks like things are much better than they were! Spreading things out more can't hurt of course though the sheer quantity of traffic may be low enough that this is just showing off(?). Any ideas on how the machine handles all this in practice? (Any packet dropping happening?)

**Comment by mundya** [comment]
As discussed offline with @neworderofjamie this is ready for merge.

**Pull request merged**